File: csharp-kata/KarateChop/src/library/KarateChop.cs
// Kata from here: http://codekata.com/kata/kata02-karate-chop/
// KarateChop a.k.a. BinarySearch

using System.Collections.Immutable;
using System.Linq;

namespace Search
{
    public class KarateChop
    {
        public static int Chop(int searchedNumber, ImmutableArray<int> arrayOfNumbers)
        {
            if (arrayOfNumbers.IsEmpty) return -1;
            
            var middleIndex = arrayOfNumbers.Length / 2;
            var middleItem = arrayOfNumbers[middleIndex];
            
            if (middleItem == searchedNumber)
            {
                return middleIndex;
            }
            else if (searchedNumber < middleItem)
            {
                return Chop(searchedNumber, arrayOfNumbers.Take(middleIndex).ToImmutableArray());
            }
            else
            {
                var index = Chop(searchedNumber, arrayOfNumbers.Skip(middleIndex+1).ToImmutableArray());
                return index == -1 ? index : middleIndex + index + 1;
            }
        }

    }
}

File: javascript-kata/BinaryChop/src/karate_chop.js
// Kata from here: http://codekata.com/kata/kata02-karate-chop/
// KarateChop a.k.a. BinarySearch

'use strict';

exports.chop = (searchedNumber, arrayOfNumbers) => {
    if (arrayOfNumbers.length === 0) return -1;        

    const MIDDLE_INDEX = Math.floor(arrayOfNumbers.length / 2);
    const MIDDLE_ITEM = arrayOfNumbers[MIDDLE_INDEX];

    if (searchedNumber === MIDDLE_ITEM) {
        return MIDDLE_INDEX;
    } else if (searchedNumber < MIDDLE_ITEM) {
        return this.chop(searchedNumber, arrayOfNumbers.slice(0, MIDDLE_INDEX));
    } else {
        const index = this.chop(searchedNumber, arrayOfNumbers.slice(MIDDLE_INDEX + 1));
        return index === -1 ? index : MIDDLE_INDEX + index + 1;
    }
        
};